Incremental auth support

squid808 opened the pull request:
This adds support of the library to authenticate with the [```include_granted_scopes```](https://developers.google.com/identity/protocols/OAuth2InstalledApp#incrementalAuth) parameter which [does not currently exist](http://stackoverflow.com/questions/33470699/how-to-implement-incremental-authorization-with-google-api#comment54766478_33470699).

Usage is intended to be exposed to a user of the library who is already using ```GoogleWebAuthorizationBroker.AuthorizeAsync()``` by adding a single new parameter - ```includeGrantedScopes``` - which is a nullable bool and set to null by default. For example:

```
userCredentials = await GoogleWebAuthorizationBroker.AuthorizeAsync(
                    MyClientSecrets, scopes, userName,
                    System.Threading.CancellationToken.None, new DataStore(),
                    true //this will set the include_granted_scopes=true in the url
                    );
```

If the user omits this optional parameter the null is trickled down and is ignored when building the authorization request URL (thereby defaulting to false in the request). When explicitly included, the query parameter is included in the URL with the value supplied, even if false.

I have tested it up to the point of verifying that it creates the url with the appropriate query parameters as expected, and ensured that the URL resolves and supplies an appropriate authentication page in the browser. I have not been able to test the resulting tokens however due to my applications being all installed applications at this time, which do not work with [Incremental Authorization](http://stackoverflow.com/questions/33514722/how-to-supply-scopes-as-part-of-post-body-rather-than-url-params/33516367?noredirect=1#comment54831247_33516367) to my knowledge.

A proper test could likely be set up to ensure that the resulting url contains the parameter when expected, however I haven't fully wrapped my head around the mocking setup used in the project's tests. If someone would be willing to guide me through it I'd be more than happy to work on that as well, though my experience with tests is somewhat limited.

This also fixes two minor typos I came across for which I saw no need for a separate pull request. It's also my first pull request, so please let me know if I have done something incorrectly during the process.

squid808 commented:
I have signed the CLA.

squid808 commented:
I have again signed the CLA with my other account I didn't realize I was using, apologies.

LindaLawton commented:
Wouldn't it be better to add it here then to just overload the method? 
[GoogleAuthorizationCodeRequestUrl.cs](https://github.com/google/google-api-dotnet-client/blob/9618f6846b29d4b812ba61362215a2c9d25110a9/Src/GoogleApis.Auth/OAuth2/Requests/GoogleAuthorizationCodeRequestUrl.cs)

peleyal commented:
Thank you very much for this pull request! It's exciting to see contribution like you trying to make this library better :+1: 

1. Before starting the review itself, I'm trying to understand what is the exact flow.
So you create a service and call the authorization method to get your credential:

userCredentials = await GoogleWebAuthorizationBroker.AuthorizeAsync("scope 1", ... , "include_granted_scopes=true").

Then you need to actually add "scope 2", since your are trying call a service that requires it.
What do you do then? How do you add "scope2" with the current code?
Should you just call to

userCredentials = await GoogleWebAuthorizationBroker.AuthorizeAsync("scope 2", ... , "include_granted_scopes=true")

and since the user already added scope1, the output credential will result in token for both "scope 1" and "scope 2"?

2. I think that Incremental Authorization should work for Installed App as well:
https://developers.google.com/identity/protocols/OAuth2InstalledApp#incrementalAuth
I will be happy to hear that it works for you, and I might try it myself after that. Thanks :)

3. Regarding testing. I can guide you, but I prefer first to focus on 1&2. Later on we can add the tests as well.

Thanks again!
Hope that my general comments seem reasonable to you.
Eyal

squid808 commented:
Linda: I'm not entirely understanding your question. I did add it to GoogleAuthorizationCodeRequestUrl.cs, but in addition I overloaded the methods in order to expose this setting easily to users of the library such as myself. Since the majority of the library tends to be abstracted away from those such as myself, after reviewing the code I determined the best way to grant a user access to control this setting was to overload the methods, starting with ```GoogleWebAuthorizationBroker.AuthorizeAsync(...)``` since that (to my understanding) is the highest level and easiest way to start with the code. Overloading the method it calls in turn also allows for catching the same functionality on multiple levels.

If I had just added it to the GoogleAuthorizationCodeRequestUrl.cs, I didn't see a good way to allow a general user to be able to append scopes, which I felt is a fairly necessary thing to be able to access and control (as opposed to ApprovalPrompt for instance, which may not be accessed or changed as often). I could, of course, have missed something in the code when I was walking through with the debugger. I tried to identify the best entry point for this code without breaking backward compatibility, so if I missed something I have no qualms against going a different direction!

Eyal, if I'm understanding your question correctly - the initial request (again, to my understanding of the api and the documentation) - wouldn't matter _what_ you passed for the parameter since there is no previous authorization to work against. In my testing, setting it to true for the first authorization didn't error out, so it seems to just ignore it in that case (since there is nothing to append scopes _to_). In the subsequent authorizations when you don't want the default behavior of replacing scopes, you do have to explicitly set the parameter to true. Setting it to false or leaving it blank (null) will induce the default behavior of replacing scopes, rather than appending.

_and since the user already added scope1, the output credential will result in token for both "scope 1" and "scope 2"?_

Based on the description [here](https://developers.google.com/identity/protocols/OAuth2InstalledApp), it will output a single token but instead of this token just representing the scopes you supplied with this auth request, it will append those scopes to the scopes you had assigned to this project from prior requests.

_I think that Incremental Authorization should work for Installed App as well:
https://developers.google.com/identity/protocols/OAuth2InstalledApp#incrementalAuth
I will be happy to hear that it works for you, and I might try it myself after that. Thanks :)_

Unfortunately, I and other users have had no such luck. At this time I don't have experience with non-installed apps to test this with but I wanted to get the ball rolling. I'd be happy to take this offline for another discussion if you feel it's not appropriate here, but until I can figure out how to do something other than an installed app without access to a webserver or something my hands are a bit tied (and no time, friggin babies man).

I'll be on the road for the next few days, so I apologize if I misread your questions in my haste or don't get back to any questions or comments soon enough! Thanks for your consideration.

peleyal commented:
Thanks for all the content you added.

I'm still not convinced that your change will actually do something, and here is the reason:

Whenever you call AuthorizeAsync, you actually call this AuthorizeAsync of the InstalledApp class:
https://github.com/google/google-api-dotnet-client/blob/9618f6846b29d4b812ba61362215a2c9d25110a9/Src/GoogleApis.Auth/OAuth2/AuthorizationCodeInstalledApp.cs#L60

By taking a look in this implementation, I see that whenever we have a token, we just use it. So I think that in your flow although we will set include_granted_scopes=true, nothing will happen. We will load the token form the data store, and because with the current code we don't compare scopes (we actually don't have a way to do it right now), we will just use it.

We need to add some path in the code to check if include_granted_scopes is set to true, and then don't use the current token.

We definitely need to test it, and make sure it works. I'm currently swamped with other things on my plate (like everyone else, right?), but I'll try to get into it later on this week.

Thanks again, hope that my comment is clear and correct me if I'm wrong regarding anything I said...
Eyal

LindaLawton commented:
Thinking out loud. 

We need a way to know what scopes the initial datastore was created with.  Then we can detect if the new scopes are being passed are the same as the old dataStore or not.   If they aren't then we can request access for these new scopes.     This would require a change to datastore to save the scopes also.

We could add an optionValues  to AuthorizeAsync support other items like the ones below.

Here are a few other things we might want to consider merging in with this 

approval_prompt https://github.com/google/google-api-dotnet-client/issues/524
login_hint            https://developers.google.com/identity/protocols/OpenIDConnect
state                 https://developers.google.com/identity/protocols/OpenIDConnect
TokenInfo validation
access_type
id_token	A JWT that contains identity information about the user that is digitally signed by Google.

peleyal commented:
I don't agree that we need to change the datastore structure, since in a
regular call to a google apis service we don't use scopes.

IMO, we just need to change the code I mentioned that if the incremental
authentication flag is true, we actually don't try to get a token from the
datastore since the token is missing scopes that we want. Then we send the
request to the authorization server using the broker we already have, get
the token, and override the existing one.


Eyal


On Mon, Nov 9, 2015, 6:49 AM Linda Lawton <notifications@github.com> wrote:

Thinking out loud.

We need a way to know what scopes the initial datastore was created with.
Then we can detect if the new scopes are being passed are the same as the
old dataStore or not. If they aren't then we can request access for these
new scopes. This would require a change to datastore to save the scopes
also.

We could add an optionValues to AuthorizeAsync support other items like the
ones below.

Here are a few other things we might want to consider merging in with this

approval_prompt #524
<https://github.com/google/google-api-dotnet-client/issues/524>
login_hint https
<https://developers.google.com/identity/protocols/OpenIDConnect>://
<https://developers.google.com/identity/protocols/OpenIDConnect>
developers.google.com
<https://developers.google.com/identity/protocols/OpenIDConnect>
/identity/protocols/
<https://developers.google.com/identity/protocols/OpenIDConnect>
OpenIDConnect
<https://developers.google.com/identity/protocols/OpenIDConnect>
state https <https://developers.google.com/identity/protocols/OpenIDConnect>
:// <https://developers.google.com/identity/protocols/OpenIDConnect>
developers.google.com
<https://developers.google.com/identity/protocols/OpenIDConnect>
/identity/protocols/
<https://developers.google.com/identity/protocols/OpenIDConnect>
OpenIDConnect
<https://developers.google.com/identity/protocols/OpenIDConnect>
TokenInfo validation
access_type
id_token A JWT that contains identity information about the user that is
digitally signed by Google.

—
Reply to this email directly or view it on
<https://github.com/google/google-api-dotnet-client/pull/624#issuecomment-155041697>
GitHub
<https://github.com/google/google-api-dotnet-client/pull/624#issuecomment-155041697>.<img
src="
https://ci6.googleusercontent.com/proxy/D5uNa2RmVkxYTTd1lcQAz_k5hQxxeVE_zYZBUWgiEv8J4XPp4y3cRWwIO80pR88zbnSZ0mtJKR0TfGb5AcESXOt7JKoAyMbhcAlm2hfpH1cpup0XuxyFTdjYBVnBHXMRMbJwuf05BWsCVEuhsP6xH9S7sqSIug=s0-d-e1-ft#https://github.com/notifications/beacon/AC8KeeAy4fqCKvuNwxRgSIwUYSVM7U54ks5pEH_ggaJpZM4GcEg2.gif
">

squid808 commented:
Understood, thanks for clearing up what you meant. I had previously played with this method until realizing I could simplify things, but I had neglected to realize what you mention, thanks for catching that. I agree that there shouldn't be a need to keep track of what scopes we're using - we just in some way need to make sure the token coming back should or shouldn't replace what is there.

I think if we update ```AuthorizeAsync()``` method to also accept a new [defaulted] options parameter, we'd not mess up any backwards compatibility, not have to worry about changing object types or casting, and wouldn't add a lot of overhead. Like she said, we could potentially add in other parameters also if they needed to be specified and handled in much the same way. In that case, rather than supplying it via the Initializer and adding it to the flow, we could just declare it in the Broker and pass it along to the ```AuthorizeAsync``` method. It would also make things easier to update for future options.

Otherwise, if we wanted to explicitly check/use these parameters via my current implementation (via the initializer -> flow path) we'd have to at least cast the Flow to access these parameters that don't exist on the interface, which is ugly.

I get back later this week, I can throw something together and see if I can get something working. I'll also look in to the other parameters Linda mentioned - if most of them don't need to be considered in ```AuthorizeAsync()``` they'll be relatively simple to add in the same way.

peleyal commented:
Thanks.
An option is to add more parameters to the initializer at:
https://github.com/google/google-api-dotnet-client/blob/master/Src/GoogleApis.Auth/OAuth2/Flows/AuthorizationCodeFlow.cs
and make this function public
https://github.com/google/google-api-dotnet-client/blob/master/Src/GoogleApis.Auth.DotNet4/OAuth2/GoogleWebAuthorizationBroker.cs#L116

(Just something to consider)

Eyal

LindaLawton commented:
I wonder if adding request_visible_actions could be added to this as well.   Posting moments to g+ still doesn't work without that.

squid808 commented:
Sorry for the delay.

> An option is to add more parameters to the initializer at:
https://github.com/google/google-api-dotnet-client/blob/master/Src/GoogleApis.Auth/OAuth2/Flows/AuthorizationCodeFlow.cs

Why add it to AuthorizationCodeFlow.Initializer instead of the GoogleAuthorizationCodeFlow.Initializer that inherits it? With this and all the other parameters suggested here, if they're going to be specific to Google as opposed to generic OAuth 2.0 (for instance, ```request_visible_actions``` which is a G+ thing) shouldn't they be in the latter?

Either way, that's the direction I'm heading - adding the parameters to the initializer and making the ```AuthorizeAsyncCore()``` public so that rather than adding parameters to the methods, they can just call it directly.

As to what I'm thinking beyond that - since we need the [```AuthorizeAsync()```](https://github.com/google/google-api-dotnet-client/blob/9618f6846b29d4b812ba61362215a2c9d25110a9/Src/GoogleApis.Auth/OAuth2/AuthorizationCodeInstalledApp.cs#L60) method to know when it needs to force the data store to update, and since it uses Flow which has a contract with [```IAuthorizationCodeFlow```](https://github.com/google/google-api-dotnet-client/blob/master/Src/GoogleApis.Auth/OAuth2/Flows/IAuthorizationCodeFlow.cs#L29) as opposed to the derived specific types, if we added a simple method of ShouldForceRefresh() that returns a ```bool```, we could have a simple ```return false;``` virtual method defined in the ```AuthorizationCodeFlow``` and then a more complex version overridden in the ```GoogleAuthorizationCodeFlow``` that checks for things like a value of true for the ```IncludeGrantedScopes``` property (and anything else in the future that might similarly affect it) and returns true in those cases.

A lot of word salad, I know. I'm working on the updates to the code now, just trying to properly add in the parameters suggested by Linda before updating the pull request.

squid808 commented:
Just a thought in a different direction and please let me know if you'd rather not have the project act this way - but what if instead of explicitly adding all these query parameters, I were to instead add an array of KeyValuePair objects that could represent any other query parameters the user specified? The user would specify that array in much the same way as we discussed - via the auth initializer - but it would give them the freedom to add in any other parameters (and as many as) they wanted. 

I've already targeted how to do it - roughly by adding a new type to the RequestParameterType enum (CustomQuery), updating the ```RequestBuilder.AddParameter()``` switch to handle that new type (and simply add any key value pairs to the QueryParameters list) and adding that property to GoogleAuthorizationCodeRequestUrl. I'm pretty sure it would work.

Assuming that most of the OpenID parameters mentioned above and other future parameters don't require an existing token to be replaced (as does the incremental auth), all it would need to do is show up in the URL, correct? The resulting code/token exchange would then proceed as normal. If so, I don't think that's in the scope of this pull request and I can open a separate one for that once I'm done with this. Please let me know your thoughts - otherwise, I'll look in to manually and explicitly adding the missing OpenID parameters as previously planned.

squid808 commented:
Linda - I left out the other parameters for now pending your thoughts on my above comment.

Eyal - I've updated the interface to allow for the Flow to indicate if/when the token should be forced to be retrieved and stored, it should alleviate the issue you described. I elected to update the Interface to keep things clean, and the method I added could have applications down the line. If you'd rather me not alter the interface I can work around it though it might not be as clean.

LindaLawton commented:
I know that request_visible_actions is proably just a G+ thing at this time but iwth out the ablity to set it we moments.insert wont work.   I have had to hack my own version of the library to get this to work.  [Wall of text on this](https://plus.google.com/+GerwinSturm/posts/Uv1JirDUE5J)

My thoughts are you are awesome i have wanted to do this for months but could never figure out how to do it and make it look "nice"  Keep up the good work.   I will leave code review to Eyal :)

peleyal commented on Src/GoogleApis.Auth.DotNet4/OAuth2/GoogleWebAuthorizationBroker.cs:
```diff
@@ -113,17 +113,22 @@ public class GoogleWebAuthorizationBroker
         /// <param name="dataStore">The data store, if not specified a file data store will be used.</param>
         /// <returns>User credential.</returns>
-        private static async Task<UserCredential> AuthorizeAsyncCore(
+        public static async Task<UserCredential> AuthorizeAsyncCore(
```
I would suggest just changing this function to be public. That's all.
Users who use the new flow will have to use this method, and set initializer.IncludeGrantedScopes before the call.
What do you think? I think it's neater.

squid808 replied:
The function of ```AuthorizeAsyncCore()``` should already be public in the latest commit unless I'm missing something, I'll double check that it is. I agree that it's neater, would you prefer to force them to use this method specifically by removing the ```includeGrantedScopes``` parameter from the ```AuthorizeAsync()```? That way they'd have no choice but to use the -Core method.

peleyal replied:
Yeah. It's not a common case, that's why I'm ok with not exposing it anywhere besides this method.
By the way, we should rename it to AuthorizeAsync (no core), not that it's public.

squid808 replied:
Understood, and done. I misinterpreted your original comment, my bad, again thanks for clarifying. I removed -Core from the method name. Not that it came up but I don't think the method summary needs changing, it's still accurate since nothing there changed other than renaming and being made public.

peleyal commented:
1. You are right, GoogleAuthorizationCodeFlow is the right place to do that. Thanks
2. I think that your idea of ShouldForceRefresh is nice. I'm still trying to figure out if it's the right way.
3. Your idea of KeyValuePair sounds like a good solution for future new parameters. I would focus for now for adding this new boolean you just created, and open a new issue \ PR for later on to add this capability as well. To be honest, I think it's a good idea for authentication and to service calls as well.
But, again, lets focus on this PR for now.

I'm starting the review now...

peleyal commented on Src/GoogleApis.Auth.DotNet4/OAuth2/GoogleWebAuthorizationBroker.cs:
```diff
@@ -113,17 +113,22 @@ public class GoogleWebAuthorizationBroker
             GoogleAuthorizationCodeFlow.Initializer initializer, IEnumerable<string> scopes, string user,
-            CancellationToken taskCancellationToken, IDataStore dataStore = null)
+            CancellationToken taskCancellationToken, IDataStore dataStore = null,
+            bool? includeGrantedScopes = null)
```
no need to include this parameter, it exists already in the initializer.

squid808 replied:
Makes sense, I was attempting to stick with the existing setup. For instance, scopes and datastore, which are also part of the initializer, are included. I wasn't sure if that was intentional or for legacy code so I erred on the side of intentional. I agree that it'd be cleaner to not have it there, and will remove it.

peleyal commented on Src/GoogleApis.Auth/OAuth2/AuthorizationCodeInstalledApp.cs:
```diff
@@ -64,7 +64,7 @@ public ICodeReceiver CodeReceiver
             // If the stored token is null or it doesn't have a refresh token and the access token is expired we need 
             // to retrieve a new authorization code.
-            if (token == null || (token.RefreshToken == null && token.IsExpired(flow.Clock)))
+            if (token == null || Flow.ShouldForceTokenRetrieval() || (token.RefreshToken == null && token.IsExpired(flow.Clock)))
```
I would change the order
if (Flow.ShouldForceTokenRetrieval() || token == null || ...

squid808 replied:
Done. I assumed that the token == null was most likely to be true, hence the provided order.

peleyal commented on Src/GoogleApis.Auth/OAuth2/Flows/AuthorizationCodeFlow.cs:
```diff
@@ -270,6 +270,7 @@ public virtual Task RevokeTokenAsync(string userId, string token, CancellationTo
             throw new NotImplementedException("The OAuth 2.0 protocol does not support token revocation.");
         }
 
+        public virtual bool ShouldForceTokenRetrieval() { return false; }
```
Please add new empty line after.

squid808 replied:
Confirmed.

peleyal commented on Src/GoogleApis.Auth/OAuth2/Flows/GoogleAuthorizationCodeFlow.cs:
```diff
@@ -35,11 +35,15 @@ public class GoogleAuthorizationCodeFlow : AuthorizationCodeFlow
         public string RevokeTokenUrl { get { return revokeTokenUrl; } }
 
+        /// <summary>Gets or sets the include granted scopes indicator.</summary>
+        private bool? includeGrantedScopes { get; set; }
```
Do we need a nullable here? maybe adding just a bool is enough?

squid808 replied:
The reason I chose a nullable was to give the user a way to indicate if they wanted it to be true or false and also have it show up in the URI either way. The default behavior or a null indeed has the same end result of false, but if it was a standard bool and they for some reason wanted/needed the false to show up in the URI (validation checks on their end, maybe? grasping at straws, tbh) they would not be able to. I would not be opposed to having it be a standard bool since it doesn't affect the end result, I just tend to try and fit all scenarios just in case. Thoughts?

peleyal replied:
You convinced me.
null -> nothing will sent on the wire
true/false -> the value will be sent as part of the URL.

That's what you meant, right?

squid808 replied:
Correct.

peleyal replied:
I would prefer us to expose a public getter, similar to the way revokeTokenUrl is declared:

private readonly bool includeGrantedScopes;

/// <summary>Gets or sets the include granted scopes indicator.</summary>
private bool? IncludeGrantedScopes { get { return includeGrantedScopes; } }

squid808 replied:
No prob, done and references updated as appropriate.

peleyal commented on Src/GoogleApis.Auth/OAuth2/Flows/GoogleAuthorizationCodeFlow.cs:
```diff
@@ -72,12 +78,25 @@ public override AuthorizationCodeRequestUrl CreateAuthorizationCodeRequest(strin
 
+        public override bool ShouldForceTokenRetrieval() 
+        {
+            if (includeGrantedScopes.HasValue && includeGrantedScopes.Value == true)
```
Please change to: 
return includeGrantedScopes.HasValue && includeGrantedScopes.Value == true;

(if you change the includeGrantedScopes to bool (and not bool?) we can just return includeGrantedScopes; :+1:

squid808 replied:
Done, pending the bool vs bool? thoughts above. You kids and your crazy shorthanded logic =) Good catch.

peleyal commented on Src/GoogleApis.Auth/OAuth2/Flows/IAuthorizationCodeFlow.cs:
```diff
@@ -86,5 +86,11 @@ public interface IAuthorizationCodeFlow : IDisposable
+        /// <summary>
+        /// Indicates if a new token needs to be retrieved and stored regardless of normal circumstances.
+        /// </summary>
+        /// <returns></returns>
```
You can remove the empty returns tag

squid808 replied:
Done.

peleyal commented on Src/GoogleApis.Auth/OAuth2/Requests/GoogleAuthorizationCodeRequestUrl.cs:
```diff
@@ -33,7 +33,7 @@ public class GoogleAuthorizationCodeRequestUrl : AuthorizationCodeRequestUrl
 
         /// <summary>
-        /// Gets pr sets prompt for consent behavior <c>auto</c> to request auto-approval or<c>force</c> to force the 
+        /// Gets or sets prompt for consent behavior <c>auto</c> to request auto-approval or<c>force</c> to force the 
```
Thanks :+1:

peleyal commented on Src/GoogleApis.Auth/OAuth2/Requests/GoogleAuthorizationCodeRequestUrl.cs:
```diff
@@ -49,6 +49,16 @@ public class GoogleAuthorizationCodeRequestUrl : AuthorizationCodeRequestUrl
         public string LoginHint { get; set; }
 
         /// <summary>
+        /// Gets or sets the Include Granted Scopes to determine if this authorization request should use
```
include granted scopes (lower case?)

squid808 replied:
Fixed, I probably had just copied in the variable name and added spaces because I'm a special kind of lazy sometimes when it gets late...

peleyal commented on Src/GoogleApis.Auth/OAuth2/Requests/GoogleAuthorizationCodeRequestUrl.cs:
```diff
@@ -49,6 +49,16 @@ public class GoogleAuthorizationCodeRequestUrl : AuthorizationCodeRequestUrl
 
         /// <summary>
+        /// Gets or sets the Include Granted Scopes to determine if this authorization request should use
+        /// incremental authorization. If this is provided with the value <c>true</c>, and the authorization 
```
Please add a link to incremental authorization - https://developers.google.com/+/web/api/rest/oauth#incremental-auth

squid808 replied:
Removed everything starting at and following "If this is provided..." and added in
*For additional details see https://developers.google.com/+/web/api/rest/oauth#incremental-auth.*

> and a .NET tag for the link)" after incremental authorization.

Sorry, not sure I follow. Do you mean something like [this](http://stackoverflow.com/a/6960573/792032)?

peleyal replied:
I'm sorry you can just change it to:
/// Gets or sets the Include Granted Scopes to determine if this authorization request should use
/// incremental authorization (https://developers.google.com/+/web/api/rest/oauth#incremental-auth). If...

Or (the like you link you shared)
/// Gets or sets the Include Granted Scopes to determine if this authorization request should use
/// &lt;see href="https://developers.google.com/+/web/api/rest/oauth#incremental-auth"&gt;incremental authorization&lt;/&gt; (). If...

squid808 replied:
Done, see impending commit.

peleyal commented on Src/GoogleApis.Auth/OAuth2/Requests/GoogleAuthorizationCodeRequestUrl.cs:
```diff
@@ -56,6 +66,7 @@ public GoogleAuthorizationCodeRequestUrl(Uri authorizationServerUrl)
             : base(authorizationServerUrl)
         {
             AccessType = "offline";
+            IncludeGrantedScopes = null;
```
is null by default. You can remove it.

squid808 replied:
Done.

peleyal commented on Src/GoogleApis.Core/Apis/Http/HttpClientFactory.cs:
```diff
@@ -54,7 +54,7 @@ protected virtual HttpMessageHandler CreateHandler(CreateHttpClientArgs args)
             var handler = new HttpClientHandler();
 
-            // If the framework supports redirect configuration, set it to false, because ConfigurableMessageHnalder 
+            // If the framework supports redirect configuration, set it to false, because ConfigurableMessageHandler 
```
Thanks again :+1:

peleyal commented:
Hi thanks for the update.
1. Several of my comments were addressed (or you just didn't response to them...)
2. I think that you didn't commit your latest changes.

Thanks so much!
Eyal

squid808 commented:
I was holding off about the nullable thing before finishing and committing/pushing. I'll get that sorted out shortly and pushed. Thanks for taking your time to write things out!

peleyal commented on Src/GoogleApis.Auth/OAuth2/AuthorizationCodeInstalledApp.cs:
```diff
@@ -64,7 +64,7 @@ public ICodeReceiver CodeReceiver
             // If the stored token is null or it doesn't have a refresh token and the access token is expired we need 
             // to retrieve a new authorization code.
-            if (token == null || (token.RefreshToken == null && token.IsExpired(flow.Clock)))
+            if (Flow.ShouldForceTokenRetrieval() || token == null || (token.RefreshToken == null && token.IsExpired(flow.Clock)))
```
[optional] What do you think? Should we wrap this if in a function
public bool ShouldRequestAuthorizationCode()
{
  return Flow.ShouldForceTokenRetrieval() || token == null || (token.RefreshToken == null && token.IsExpired(flow.Clock));
}

I think it will be neater. If you do so, please add a comment as well.

squid808 replied:
Done. I thought far too long on this but eventually decided to throw it in a separate function to keep it a bit cleaner. I kept and adjusted the in-line comments since I figure they were there for a reason. Part of me wanted to find a way to only check for a token if ```Flow.ShouldForceTokenRetrieval()``` is false anyways, but this should be just fine for now.

peleyal commented on Src/GoogleApis.Auth/OAuth2/Flows/GoogleAuthorizationCodeFlow.cs:
```diff
@@ -48,7 +52,9 @@ public override AuthorizationCodeRequestUrl CreateAuthorizationCodeRequest(strin
                 Scope = string.Join(" ", Scopes),
-                RedirectUri = redirectUri
+                RedirectUri = redirectUri,
+                IncludeGrantedScopes = (includeGrantedScopes.HasValue)
```
What am I missing? Why not just:
IncludeGrantedScopes = includeGrantedScopes ?

peleyal replied:
You can remove the parenthesis around includeGrantedScopes.HasValue

squid808 replied:
Done.

peleyal commented on Src/GoogleApis.Auth/OAuth2/Flows/GoogleAuthorizationCodeFlow.cs:
```diff
@@ -72,12 +78,20 @@ public override AuthorizationCodeRequestUrl CreateAuthorizationCodeRequest(strin
 
+        public override bool ShouldForceTokenRetrieval() 
+        {
+            return includeGrantedScopes.HasValue && includeGrantedScopes.Value == true;
```
oops, actually you don't need the "== true" as well so:
return includeGrantedScopes.HasValue && includeGrantedScopes.Value;

squid808 replied:
Done. Sorry, I tend to write code almost as verbosely as I do when I write everything else. I'm sure you've noticed =D

peleyal commented on Src/GoogleApis.Auth/OAuth2/Requests/GoogleAuthorizationCodeRequestUrl.cs:
```diff
@@ -49,6 +49,16 @@ public class GoogleAuthorizationCodeRequestUrl : AuthorizationCodeRequestUrl
         /// <summary>
+        /// Gets or sets the include granted scopes to determine if this authorization request should use
+        /// incremental authorization (https://developers.google.com/+/web/api/rest/oauth#incremental-auth).
+        /// If this is provided with the value <c>true</c>, and the authorization request is granted, the 
```
[optional]
I think you can replace "If this is provided with the value true..." with "If true..."

squid808 replied:
Makes sense, more succinct. Done.

> I think that we can also remove the c tag, and just write "true" or "false" since it's a test property

I don't quite follow, but also done - not including the quotes I assume.

peleyal commented on Src/GoogleApis.Auth/OAuth2/Flows/GoogleAuthorizationCodeFlow.cs:
```diff
@@ -35,11 +35,15 @@ public class GoogleAuthorizationCodeFlow : AuthorizationCodeFlow
+            includeGrantedScopes = initializer.IncludeGrantedScopes;
         }
 
         public override AuthorizationCodeRequestUrl CreateAuthorizationCodeRequest(string redirectUri)
```
It will be nice to add a test to this method (currently we don't have unit tests to this class at all).

You can add it, testing a case where includeGrantedScopes is set \ unset, or just create a new issue in the issue tracker to add unit tests for this class.
Thanks!

peleyal commented:
We are very very close. I just want to verify that you actually tested your code with real app and it works as expected. If not, I think it's the right time :+1:

Good job!

squid808 commented:
Thanks again for your comments, quite helpful.

I'll take a look at unit tests to see if it's something I can throw together in the next few days. While I've been using a custom test program to manually verify the URL contains the parameters as expected, I think I'd feel more comfortable having the unit tests to back me up. It'll take me a little while to read through the other tests for examples to get a feel for how you're setting up the mock objects or whatever is needed but I'll be happy to give it a go.

I still can't verify, however, that putting it in the url actually bears fruit - i'll be finishing up that test program and throwing it on SO shortly as well. Basically, I can verify that it shows up in the URL, but not that it makes any difference to the returned code and tokens. But, since the scope of this PR is to get the query parameter in the URL, I'm trying to keep that separate. Let me know if you think my thought process on that is flawed.

squid808 commented:
Ok, I studied the other Unit Tests and I don't think I'm quite ready to figure all that out yet, sorry. I get the basics, but I get stuck on how to appropriately break down the tests and what to look for. So, instead I threw a quick test program together, it and the results [can be viewed here](https://gist.github.com/squid808/d8521f170dacc955f6c5).

The results of these code changes seem to be working as intended - supplying true or false to the parameter will cause ```include_granted_scopes``` to show up in the URL, but false does not cause the user to reauthenticate (though it does show up in the URL if it's the first authentication).

Notes on the issue with Incremental Authorization not working with an Installed App type program can be found either in the note on that or in the initial post in this PR. Again, I only bring that up here in case the issue there is actually due to this code not being implemented properly as opposed to something on the backend.

peleyal commented on Src/GoogleApis.Auth/OAuth2/AuthorizationCodeInstalledApp.cs:
```diff
@@ -88,7 +87,18 @@ public ICodeReceiver CodeReceiver
+        {
+            // If the flow includes a parameter that requires a new token, if the stored token is null or it doesn't
+            // have a refresh token and the access token is expired we need to retrieve a new authorization code.
+            return Flow.ShouldForceTokenRetrieval() || token == null || (token.RefreshToken == null && token.IsExpired(flow.Clock));
```
Just verify that the line length is no more than 120 :) It's hard to verify that using this viewer...

squid808 replied:
It was, thank you. Fixed. I broke the line before the operand of && so that a glance at the next line shows it's part of the previous line's operation, I'm not sure if there is some best practice I should otherwise be adhering to there. If you'd rather the line break be after the operand before the variable let me know.

peleyal commented on Src/GoogleApis.Auth/OAuth2/Requests/GoogleAuthorizationCodeRequestUrl.cs:
```diff
@@ -49,6 +49,16 @@ public class GoogleAuthorizationCodeRequestUrl : AuthorizationCodeRequestUrl
+        /// Gets or sets the include granted scopes to determine if this authorization request should use
+        /// incremental authorization (https://developers.google.com/+/web/api/rest/oauth#incremental-auth).
+        /// If true and the authorization request is granted, the authorization will include any previous 
+        /// authorizations granted to this user/application combination for other scopes. Default value is 
```
Default value is null, right? I would omit the last sentence.

squid808 replied:
I've removed it. I guess in my brain place I was trying to explain the resulting behavior of the URL and authorization, perhaps it would have been better as:

> Default *behavior* is false *which* will be omitted unless explicitly indicated.

But, that's a summary of the authorization behavior not the method, so I suppose it doesn't belong there.

peleyal commented:
The code is super neat.
I saw your comment, mentioning that this code doesn't actually work for installed app. Did you try to test it with a web application?
I'll try to test it myself using your app later on this week.
Thanks!

peleyal commented on Src/GoogleApis.Auth/OAuth2/AuthorizationCodeInstalledApp.cs:
```diff
@@ -88,7 +87,18 @@ public ICodeReceiver CodeReceiver
             }
 
-            return new UserCredential(flow, userId, token);
+            return new UserCredential(Flow, userId, token);
```
You can leave it as return new UserCredential(flow, userId, token); (lower f)

squid808 replied:
Fixed. I have no recollection of making this particular change, and if it weren't for the evidence in front of my face I'd swear you were trying to pull one over on me...

squid808 commented:
Thanks again, you're too kind! My programming origins are isolated and humble indeed and I know I have a lot to pick up on, so I really do appreciate how cool and patient you've been with the whole process.

Unfortunately I've only ever dealt with the installed app scenarios. I'll look around for some sample code I might be able to quickly run without having to do a ton of binding redirects and modifications to in order to get it running. I'll check out the quick-start app mentioned [here](https://developers.google.com/%2B/web/samples/csharp?hl=en) later when I can keep my eyes open a bit longer =)

peleyal commented:
Hi thanks again :) You're pretty awesome as well.
I just ran an installed app with two different services, and, like you mentioned, the original scope was erased.
I'm going to contact some developers form the auth group to understand why it happens.

squid808 commented:
Thanks so much! That means there is a light at the end of the tunnel for my users and me!

peleyal commented:
There is always light!
Hopefully it's something that we are doing wrong, so we will be able to fix it soon.

I'll keep you updated,
Eyal

LindaLawton commented:
@squid808  and i went though this a little on stack all the documentation we found appears to refer only to web and mobile apps.    It appear it wasn't implemented for installed apps.  

I can put together a quick test without the client library to find out if it works or not.   Either way look forward to hearing what the Oauth gurus say.

peleyal commented:
Someone also mentioned to me that Incremental Auth doesn't work for installed app.
1. We need to test an ASP.NET case, or WP...
If it works, 
2. It will be a good idea to update our docs

Then we can commit :+1: 

I'll try to spend sometime with it later on this week. @squid808 if you have time, you should test it too.
Thanks,
Eyal

peleyal commented:
OK. I tested your code and with a small fix it works for **Web applications**

1. We need to change
https://github.com/google/google-api-dotnet-client/blob/master/Src/GoogleApis.Auth/OAuth2/Web/AuthorizationCodeWebApp.cs#L105
as well (as similar to the change you had for the Installed App

2. We need to add some documentation about the fact that include grants wouldn't work for installed app, since it isn't support yet by the backend (as of November 2015).

It might be changed in the future but for now, it isn't supported. An alternative is to delete your code form the Installed Apps flow, but I think that we should leave it as it is, because maybe in the near future this support will be added.

I would add explanation that Incremental Auth doesn't work for installed application, for now...

https://github.com/squid808/google-api-dotnet-client/blob/IncrementalAuthSupport/Src/GoogleApis.Auth/OAuth2/Requests/GoogleAuthorizationCodeRequestUrl.cs#L59
and to the installed app class documentation:
https://github.com/squid808/google-api-dotnet-client/blob/IncrementalAuthSupport/Src/GoogleApis.Auth/OAuth2/Flows/IAuthorizationCodeFlow.cs

3. We should hurry and add those changes before Monday, because we are working on creating a new release (1.10) in the next few days.
Otherwise, it will be part of 1.11

4. Thanks!

squid808 commented:
Sorry - I spent some time trying to learn how to do web dev enough to refactor the task api sample to work, but I hadn't gotten far enough in the limited time I had. You beat me to it =). I did however see that we needed to add it to the web auth, so I had done that. However, I duplicated the ShouldRequestAuthorizationCode for now to get things working similarly. I assume it should go in some utility class, but I wasn't sure if you had a particular class in mind.

That said, I added a sentence description for each class you mentioned (I assumed your second link was a typo and you actually meant the AuthorizationCodeInstalledApp.cs file?), let me know if it's good or not. I'll check in periodically tomorrow to try and get this done.

peleyal commented on Src/GoogleApis.Auth/OAuth2/AuthorizationCodeInstalledApp.cs:
```diff
@@ -62,9 +63,8 @@ public ICodeReceiver CodeReceiver
-            // If the stored token is null or it doesn't have a refresh token and the access token is expired we need 
-            // to retrieve a new authorization code.
-            if (token == null || (token.RefreshToken == null && token.IsExpired(flow.Clock)))
+            //Check if a new authorization code is needed.
```
Please add space before Check

peleyal commented on Src/GoogleApis.Auth/OAuth2/Web/AuthorizationCodeWebApp.cs:
```diff
@@ -102,7 +103,7 @@ public AuthorizationCodeWebApp(IAuthorizationCodeFlow flow, string redirectUri,
 
             // If the stored token is null or it doesn't have a refresh token and the access token is expired, we need 
```
Change the comment with this one:
// Check if a new authorization code is needed.

peleyal commented:
Thanks for the quick response. Just fix the small comments issues I raised, and we are ready to submit it.
Thanks so much for being patient and thorough!

squid808 commented:
Ok, should be good per your comments. I noticed a spacing issue just after I hit push, and thus the second. Sorry for all the extra work, I feel like you've written more here in the comments than you would have just adding the code yourself. I'll check back throughout today in case you spot something else I missed. Thanks!

peleyal commented on Src/GoogleApis.Auth/OAuth2/Web/AuthorizationCodeWebApp.cs:
```diff
@@ -19,6 +19,7 @@
 using System.Threading.Tasks;
 
 using Google.Apis.Auth.OAuth2.Flows;
+using Google.Apis.Auth.OAuth2.Responses;
```
Please change the order, it should come after 
using Google.Apis.Auth.OAuth2.Requests;

peleyal commented on Src/GoogleApis.Auth/OAuth2/Requests/GoogleAuthorizationCodeRequestUrl.cs:
```diff
@@ -49,6 +49,17 @@ public class GoogleAuthorizationCodeRequestUrl : AuthorizationCodeRequestUrl
+        /// If true and the authorization request is granted, the authorization will include any previous 
+        /// authorizations granted to this user/application combination for other scopes.
+        /// </summary>
+        /// <remarks> Currently unsupported for installed apps. </remarks>
```
No need in space before and after your sentence. Please remove the extra space before "Currently" and after "apps."

peleyal commented:
I didn't. That's the process of the review. I added two tiny comments, address them and we are ready to push :+1: That's great because it's going to be part of the coming release! Exciting :)

squid808 commented:
Done and done, thanks!

peleyal commented:
I just thought to mention here the following SO thread:
http://stackoverflow.com/questions/33982392/specify-login-hint-using-net-google-apis-oauth2-v2
@squid808 you suggested in the past supporting a generic way to add any URL parameter. Do you want to create a new issue for it and tackle it in the next few weeks?

Thanks!
Eyal

squid808 commented:
I actually have it mostly done and seemingly working - I've been keeping it in a separate branch until we merged this one. I can create a separate issue if you'd like, or just submit the pull request once it's ready. I figured this one was a little more simple to cut my teeth on for the unit tests, so I was attempting that - and I'm also trying to figure out how to do the web application to test it in. Now that I'm a little more familiar with the process, I can try to get as much of that done first as I can.

I also have to create the issue to request the unit tests for ```GoogleAuthorizationCodeFlow``` at the least, as we had discussed.

Either way, your call - I can create an issue for both if you'd prefer.

peleyal commented:
1. I think we should have a new Issue & PR for supporting any URL parameters, just to make things clearer.
2. For testing your change in ASP.NET I hacked the following sample: https://github.com/google/google-api-dotnet-client-samples/tree/master/Tasks.SimpleOAuth2
3. Yes, please add a new issue for adding unit tests for GoogleAuthorizationCodeFlow.

Thanks!
Eyal
